src/services/cipher.service.ts

```diff
@@ -1002,4 +1002,30 @@ export class CipherService implements CipherServiceAbstraction {
                 throw new Error('Unknown cipher type.');
         }
     }
+
+    private async getCipherForUrl(url: string, lastUsed: boolean): Promise<CipherView> {
+        const ciphers = await this.getAllDecryptedForUrl(url);
+        if (ciphers.length === 0) {
+            return null;
+        }
+
+        ciphers.sort(this.sortCiphersByLastUsed);
+
+        const lastUsedCipher = ciphers[0];
+
+        if (lastUsed) {
+            return lastUsedCipher;
+        }
+
+        if (!this.arraysContainSameElements(this.lastSortedCiphers, ciphers)) {
```

Comment:
Okay, didn't catch this before, however.... I've walked through the logic and perhaps I'm missing something, but it would seem that `this.lastSortedCiphers` is not necessary since you're still getting all decrypted for the URL, then always sorting them, so you're never really "using" the "cache" of prior sorted ciphers and adding a bit of overhead to determine if there were changes between the prior and current list, then update it, but you could just always use the local const variable, `ciphers`, because that's the only scope both are used.

Reply:
I would also assume some invalidation of this approach (or worse a race-condition/thread overstepping) as well when there are multiple tabs open in the browser but only a singleton instance of the cipher service itself.

Reply:
`this.lastSortedCiphers` gets sorted and populated only once (for URL/tab) and it maintains the original sorting order for every keystroke. If I'd remove it and use the local sorted array then you would cycle through the same two items all the time, because that array is always sorted by last used. Imagine 4 ciphers: 

[A, B, C, D]

*Autofill command pressed*

`const ciphers = [A, B, C, D]`
`this.lastSortedCiphers = [A, B, C, D]`

**Returns B (lastUsedIndex + 1)**

B is now the last used

*Autofill command pressed*

`const ciphers = [B, A, C, D] // Changed`
`this.lastSortedCiphers = [A, B, C, D] // Didn't change because arraysContainSameElements == true`

**Returns C (lastUsedIndex + 1)**

C is now the last used

So far, everything is the same, but if you press again the shortcut:

*Autofill command pressed*

`const ciphers = [C, B, A, D] // Changed`
`this.lastSortedCiphers = [A, B, C, D] // Didn't change`

There. Last used was C, so if we use local ciphers array, next one would be B again, but it should be D.

If you run it again, you will get C, and then B, and then C... 🙂


P.S. Writing this example, I realised that in the first step, you get B, because is the next one. But what if you have 'Auto fill on page load' disabled? Then it should autofill the last used, not the next one. I'll see what I can think of to solve than scenario.

Reply:
> I would also assume some invalidation of this approach (or worse a race-condition/thread overstepping) as well when there are multiple tabs open in the browser but only a singleton instance of the cipher service itself.

As for this, you are totally right. Although you'd have to use autofill shortcut in two tabs at the same time, wouldn't you? Anyway, I'll see what I can do. Do you have any suggestion?

Reply:
>Although you'd have to use autofill shortcut in two tabs at the same time, wouldn't you?

yes, assuming so, which seems far-fetched, however I would also assume if you had auto-fill on page load enabled and popped open several tabs at once or in short order, and they loaded at variable times in an interweaving manner... I would think you could get around this by using a hashtable/object based on the URL, however that could lead to memory bloat if there's no routine for cleanup. Perhaps it's not a concern at the moment (until if/when it becomes a concern)

>But what if you have 'Auto fill on page load' disabled? Then it should autofill the last used, not the next one. I'll see what I can think of to solve than scenario.

I would assume when auto fill on page load is **disabled** that is when your changes are appropriate, correct? You would always want to use the "next" one (agree the initial state should return index `0`, not `0 + 1`),

If auto fill on page load is **enabled** I believe should be handled in the `browser` extension's code itself in that scenario when this service is called to activate auto-fill on initial page load (being implementation specific).

Reply:
> however I would also assume if you had auto-fill on page load enabled and popped open several tabs at once or in short order, and they loaded at variable times in an interweaving manner...

No, because autofill on page load will return earlier:
```javascript
        if (lastUsed) {
            return lastUsedCipher;
        }
```
Because it uses `getLastUsedForUrl` method instead of `getNextCipherForUrl`.


> I would assume when auto fill on page load is **disabled** that is when your changes are appropriate, correct? You would always want to use the "next" one (agree the initial state should return index `0`, not `0 + 1`),
> 
> If auto fill on page load is **enabled** I believe should be handled in the `browser` extension's code itself in that scenario when this service is called to activate auto-fill on initial page load (being implementation specific).

I just remembered why I needed that `this.lastUsedCipher` in the previous PR 🤣 So if it was filled, it's because an autofill already happened (either on page load or shortcut triggered), so you would get next cipher in line (sorted by date). If not, you will get last used (position 0).

That's what happens when you pick up a PR from a year ago 😄

Reply:
haha, for sure, good catch and glad you caught that 😉